Feature: File transfer in secure conversations

vdveer opened the pull request:
* Add files to secure conversations by selecting in app (after enabling the preference)
* Share files (up to mms-sizelimit) and attach them to a secure chat

vdveer commented:
@BastienLQ : I have moved things here, since the attachment refactor on the upstream made me create a new branch to work with.

vdveer commented:
LABEL : donotmerge (i need to change the contenttype to a vendor-specific type)

pR0Ps commented on src/org/smssecure/smssecure/ConversationActivity.java:
```diff
@@ -421,7 +427,7 @@ public void onMuted(final long until) {
           @Override
           protected Void doInBackground(Void... params) {
             DatabaseFactory.getRecipientPreferenceDatabase(ConversationActivity.this)
-                           .setMuted(recipients, until);
```
Please avoid touching code unrelated to PR

pR0Ps commented on src/org/smssecure/smssecure/ConversationActivity.java:
```diff
@@ -507,7 +513,7 @@ private void handleStartSecureSession() {
     builder.setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
       @Override
       public void onClick(DialogInterface dialog, int which) {
-        if (!isEncryptedConversation){
```
same

pR0Ps commented on src/org/smssecure/smssecure/ConversationActivity.java:
```diff
@@ -516,7 +522,7 @@ public void onClick(DialogInterface dialog, int which) {
           allocatedThreadId = threadId;
         }
-        Log.w(TAG, "Refreshing thread "+allocatedThreadId+"...");
+        Log.w(TAG, "Refreshing thread " + allocatedThreadId + "...");
```
same

pR0Ps commented on src/org/smssecure/smssecure/ConversationActivity.java:
```diff
@@ -681,10 +687,20 @@ private void handleManualMmsRequired() {
+    if (draftMedia != null && draftMediaType != null) {
+      Recipient    primaryRecipient       = getRecipients() == null ? null : getRecipients().getPrimaryRecipient();
+      boolean      isSecureSmsDestination = isSingleConversation() && SessionUtil.hasSession(this, masterSecret, primaryRecipient);
+      if(draftMediaType.equals(MediaType.FILE) && !isSecureSmsDestination){
```
space between `if` and `(`

pR0Ps commented on src/org/smssecure/smssecure/ConversationActivity.java:
```diff
@@ -681,10 +687,20 @@ private void handleManualMmsRequired() {
+      if(draftMediaType.equals(MediaType.FILE) && !isSecureSmsDestination){
+        Toast.makeText(this, R.string.ConversationActivity_sorry_no_files_attaching_on_insecure_chat, Toast.LENGTH_LONG).show();
+      } else {
+        if(draftFilename != null) setMedia(draftMedia, draftMediaType, draftFilename);
```
Can pull into the parent `if` as an `else if`

pR0Ps commented on src/org/smssecure/smssecure/ConversationActivity.java:
```diff
@@ -723,7 +739,10 @@ protected void onPostExecute(List<Draft> drafts) {
+          } else if (draft.getType().equals(Draft.FILE)) {
+            setMedia(Uri.parse(draft.getValue()) , MediaType.FILE);
           }
+
```
remove whitespace

pR0Ps commented on src/org/smssecure/smssecure/ConversationActivity.java:
```diff
@@ -966,16 +988,24 @@ public void onReceive(Context context, Intent intent) {
-      AttachmentManager.selectContactInfo(this, PICK_CONTACT_INFO); break;
-    case AttachmentTypeSelectorAdapter.TAKE_PHOTO:
-      attachmentManager.capturePhoto(this, TAKE_PHOTO); break;
+      case AttachmentTypeSelectorAdapter.ADD_IMAGE:
```
Please keep previous formatting

pR0Ps commented on src/org/smssecure/smssecure/ConversationItem.java:
```diff
@@ -525,7 +529,16 @@ public void onClick(final View v, final Slide slide) {
+          }
+        });
+
+      }else{
```
`} else {`

pR0Ps commented on src/org/smssecure/smssecure/ConversationItem.java:
```diff
@@ -525,7 +529,16 @@ public void onClick(final View v, final Slide slide) {
 
         context.startActivity(intent);
-      } else {
+      } else if (slide.hasFile()){
```
...`hasfile()) {`

pR0Ps commented on src/org/smssecure/smssecure/attachments/Attachment.java:
```diff
@@ -85,4 +87,7 @@ public void setThumbnail(@Nullable Bitmap thumbnail) {
   }
+
+  @Nullable
+  public String getFileName(){ return fileName; }
```
```java
public String getFileName() {
  return fileName; 
}
```

pR0Ps commented on src/org/smssecure/smssecure/attachments/UriAttachment.java:
```diff
@@ -49,4 +56,28 @@ public boolean equals(Object other) {
+      }
+    }
+    return fileName;
+    }
```
unindent 1 level

pR0Ps commented on src/org/smssecure/smssecure/database/AttachmentDatabase.java:
```diff
@@ -296,6 +298,7 @@ void insertAttachmentsForMessage(@NonNull MasterSecret masterSecret,
     DatabaseAttachment databaseAttachment = (DatabaseAttachment) attachment;
     File               dataFile           = getAttachmentDataFile(databaseAttachment.getAttachmentId(), DATA);
 
+
```
remove whitespace

pR0Ps commented on src/org/smssecure/smssecure/database/AttachmentDatabase.java:
```diff
@@ -305,6 +308,8 @@ void insertAttachmentsForMessage(@NonNull MasterSecret masterSecret,
     ContentValues contentValues = new ContentValues();
     contentValues.put(SIZE, dataSize);
     contentValues.put(CONTENT_TYPE, mediaStream.getMimeType());
+    if(attachment.getFileName() != null)
```
Add space between `if` and `(``, use brackets

pR0Ps commented on src/org/smssecure/smssecure/database/AttachmentDatabase.java:
```diff
@@ -463,6 +470,8 @@ private AttachmentId insertAttachment(MasterSecret masterSecret, long mmsId, Att
     contentValues.put(CONTENT_LOCATION, attachment.getLocation());
     contentValues.put(CONTENT_DISPOSITION, attachment.getKey());
     contentValues.put(NAME, attachment.getRelay());
+    if(attachment.getFileName() != null)
```
Same as above

pR0Ps commented on src/org/smssecure/smssecure/database/DatabaseFactory.java:
```diff
@@ -559,8 +564,8 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
-        executeStatements(db, new String[] {
-            "CREATE INDEX IF NOT EXISTS draft_thread_index ON drafts (thread_id);",
+        executeStatements(db, new String[]{
+                "CREATE INDEX IF NOT EXISTS draft_thread_index ON drafts (thread_id);",
```
unrelated to PR

pR0Ps commented on src/org/smssecure/smssecure/database/DraftDatabase.java:
```diff
@@ -126,10 +127,11 @@ public String getValue() {
-      case IMAGE: return context.getString(R.string.DraftDatabase_Draft_image_snippet);
-      case VIDEO: return context.getString(R.string.DraftDatabase_Draft_video_snippet);
-      case AUDIO: return context.getString(R.string.DraftDatabase_Draft_audio_snippet);
+        case TEXT:  return value;
```
leave previous formatting

pR0Ps commented on src/org/smssecure/smssecure/database/MmsSmsDatabase.java:
```diff
@@ -172,12 +176,12 @@ private Cursor queryTables(String[] projection, String selection, String order,
-                                  " FROM " + AttachmentDatabase.TABLE_NAME + " WHERE " +
-                                  AttachmentDatabase.TABLE_NAME + "." + AttachmentDatabase.MMS_ID + " = " +
-                                  MmsDatabase.TABLE_NAME + "." + MmsDatabase.ID + " LIMIT 1)");
+            AttachmentDatabase.TABLE_NAME +
```
Leave previous formatting

pR0Ps commented on src/org/smssecure/smssecure/jobs/MmsSendJob.java:
```diff
@@ -234,6 +234,8 @@ private SendReq constructSendPdu(MasterSecret masterSecret, OutgoingMediaMessage
         PduPart part = new PduPart();
         part.setData(Util.readFully(PartAuthority.getAttachmentStream(context, masterSecret, attachment.getDataUri())));
         part.setContentType(Util.toIsoBytes(attachment.getContentType()));
+        if(attachment.getFileName() != null)
```
space between `if` and `(`, add brackets

pR0Ps commented on src/org/smssecure/smssecure/mms/AttachmentManager.java:
```diff
@@ -82,6 +80,24 @@ public AttachmentManager(@NonNull Activity activity, @NonNull AttachmentListener
     thumbnail.setOnClickListener(new ThumbnailClickListener());
   }
 
+  //public void clearIfFileSlides(){
```
Remove

pR0Ps commented on src/org/smssecure/smssecure/mms/AttachmentManager.java:
```diff
@@ -82,6 +80,24 @@ public AttachmentManager(@NonNull Activity activity, @NonNull AttachmentListener
+
+
+  public void selectFile(final MasterSecret secret, final Activity activity, int requestCode) {
+        new FileChooser(activity).setFileListener(new FileChooser.FileSelectedListener() {
```
correct indentation

pR0Ps commented on src/org/smssecure/smssecure/mms/AttachmentManager.java:
```diff
@@ -92,7 +108,8 @@ public void onSuccess(Boolean result) {
       }
 
       @Override
-      public void onFailure(ExecutionException e) {}
```
unrelated to PR

pR0Ps commented on src/org/smssecure/smssecure/mms/AttachmentManager.java:
```diff
@@ -153,11 +178,11 @@ public void setMedia(@NonNull final MasterSecret masterSecret,
+          final Slide slide     = mediaType.createSlide(context, uri, mediaSize, fileName);
           Log.w(TAG, "slide with size " + mediaSize + " took " + (System.currentTimeMillis() - start) + "ms");
           return slide;
-        } catch (IOException ioe) {
```
unrelated to PR

pR0Ps commented on src/org/smssecure/smssecure/mms/AttachmentManager.java:
```diff
@@ -305,19 +330,20 @@ public void onClick(View v) {
-      case VIDEO: return new VideoSlide(context, uri, dataSize);
-      default:    throw  new AssertionError("unrecognized enum");
+        case IMAGE: return new ImageSlide(context, uri, dataSize);
+        case GIF:   return new GifSlide(context, uri, dataSize);
```
keep previous indentation

pR0Ps commented on src/org/smssecure/smssecure/mms/AttachmentManager.java:
```diff
@@ -305,19 +330,20 @@ public void onClick(View v) {
-    {
+                                               long    dataSize,
+                                                String fileName)
+            throws IOException {
```
keep previous indentation

pR0Ps commented on src/org/smssecure/smssecure/mms/AttachmentTypeSelectorAdapter.java:
```diff
@@ -71,22 +75,35 @@ public View getView(int position, View convertView, ViewGroup parent) {
+    addItem(data, context.getString(R.string.AttachmentTypeSelectorAdapter_audio), ResUtil.getDrawableRes(context, R.attr.conversation_attach_sound), ADD_SOUND);
     addItem(data, context.getString(R.string.AttachmentTypeSelectorAdapter_contact), ResUtil.getDrawableRes(context, R.attr.conversation_attach_contact_info), ADD_CONTACT_INFO);
-
+      if(isEncryptedConversation && SMSSecurePreferences.getMultipartMMS(context)) {
```
space between `if` and `(`, fix indentation

pR0Ps commented on src/org/smssecure/smssecure/mms/AttachmentTypeSelectorAdapter.java:
```diff
@@ -71,22 +75,35 @@ public View getView(int position, View convertView, ViewGroup parent) {
   private static void addItem(List<IconListItem> list, String text, int resource, int id) {
     list.add(new IconListItem(text, resource, id));
   }
+    public void setSecureDestination(boolean isSecure){
```
Fix indentation

pR0Ps commented on src/org/smssecure/smssecure/mms/FileChooser.java:
```diff
@@ -0,0 +1,134 @@
+import java.io.FileFilter;
+import java.util.Arrays;
+
+public class FileChooser {
```
Make indentation in this file conform to the rest of the project

pR0Ps commented on src/org/smssecure/smssecure/mms/FileChooser.java:
```diff
@@ -0,0 +1,134 @@
+    private String extension = null;
+    public void setExtension(String extension) {
+        this.extension = (extension == null) ? null :
+                extension.toLowerCase();
```
can go on same line

pR0Ps commented on src/org/smssecure/smssecure/mms/FileSlide.java:
```diff
@@ -0,0 +1,75 @@
+import ws.com.google.android.mms.ContentType;
+import ws.com.google.android.mms.pdu.PduPart;
+
+public class FileSlide extends Slide {
```
Fix indentation for file

pR0Ps commented on src/org/smssecure/smssecure/mms/MmsMediaConstraints.java:
```diff
@@ -38,4 +38,7 @@ public int getVideoMaxSize() {
   }
+
+  @Override
+  public int getFileMaxSize() { return MAX_MESSAGE_SIZE; }
```
put function contents on a new line

pR0Ps commented on src/org/smssecure/smssecure/mms/OutgoingMediaMessage.java:
```diff
@@ -1,7 +1,9 @@
 package org.smssecure.smssecure.mms;
 
```
Seems like nothing in this file needs to change?

pR0Ps commented on src/org/smssecure/smssecure/mms/Slide.java:
```diff
@@ -38,7 +38,6 @@
   public Slide(@NonNull Context context, @NonNull Attachment attachment) {
     this.context    = context;
     this.attachment = attachment;
-
```
unrelated to PR

pR0Ps commented on src/org/smssecure/smssecure/mms/Slide.java:
```diff
@@ -72,6 +71,10 @@ public boolean hasAudio() {
     return false;
   }
 
+  public boolean hasFile() { return false; }
```
function contents on new lines

pR0Ps commented on src/org/smssecure/smssecure/mms/Slide.java:
```diff
@@ -102,10 +105,21 @@ public boolean hasPlaceholder() {
   {
     Optional<String> resolvedType = Optional.fromNullable(MediaUtil.getMimeType(context, uri));
-    return new UriAttachment(uri, resolvedType.or(defaultMime), AttachmentDatabase.TRANSFER_PROGRESS_STARTED, size);
+    if(fileName != null)
```
space between `if` and `(` and add brackets

pR0Ps commented on src/org/smssecure/smssecure/mms/Slide.java:
```diff
@@ -118,6 +132,7 @@ public boolean equals(Object other) {
            this.hasAudio() == that.hasAudio()                        &&
            this.hasImage() == that.hasImage()                        &&
            this.hasVideo() == that.hasVideo()                        &&
+           this.hasFile() == that.hasFile()                          &&
```
line up with line above

pR0Ps commented on src/org/smssecure/smssecure/mms/SlideDeck.java:
```diff
@@ -86,7 +86,29 @@ public void addSlide(Slide slide) {
+
+  public boolean removeFileSlides() {
+    boolean returnObject = false;
+    if(slides.isEmpty())
```
space and brackets

pR0Ps commented on src/org/smssecure/smssecure/util/MediaUtil.java:
```diff
@@ -88,6 +90,8 @@ public static Slide getSlideForAttachment(Context context, Attachment attachment
              ? ContentType.IMAGE_JPEG
              : mimeType;
     default:
+      if(ContentType.isNonTextVideoImageAudioType(mimeType))
```
space and brackets

pR0Ps commented:
@vdveer Thanks for this!

I didn't really have too much time to look at the actual content at the moment, but I highlighted a few coding style issues.

In general, leave things the way they are if possible and follow the coding style of the surrounding code (brackets, indentation, etc)

vdveer commented:
@pR0Ps Do you use some kind of formatting plugin or is there any way to include the formatting profile in the repository? A lot of the formatting issues where made by some form of autoformatting.

pR0Ps commented:
I typically just use Android Studi, IIRC there is an option in it to match
the existing formatting. I also make sure to just go over the diff manually
and look for formatting issues before committing it.

@BastienLQ: Do you have a different strategy?

On Wed, Mar 23, 2016 at 6:13 AM, vdveer <notifications@github.com> wrote:

> @pR0Ps <https://github.com/pR0Ps> Do you use some kind of formatting
> plugin or is there any way to include the formatting profile in the
> repository? A lot of the formatting issues where made by some form of
> autoformatting.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/SMSSecure/SMSSecure/pull/365#issuecomment-200282015>
>

BastienLQ commented:
I use Atom and it formats code automagically based on previous indentation.

BastienLQ commented:
@vdveer Can you rebase your commit on `upstream`?

vdveer commented:
@BastienLQ Sure, I will get to that a little later. In the meantime a question; when reading contribution policy for the whispersystems upstream, I read the words 'think twice when adding a user preference'. In this PR I introduced a preference to activate filetransfer. I introduced this so the user is sure what he wants when he applies the option, also because when a user sends a file to a user with an older version of the app will receive a message without file attachment (text-only when text is also added). What do you think is the best way to go here? Leave the preference or remove it and give everyone the file option in the menu on secure threads?

BastienLQ commented:
Actually, as I just merged `upstream` into `master`, can you rebase your branch on `SMSSecure:master`?
You can remove the preference you added.

vdveer commented:
@BastienLQ removed preference, rebased on master, squashed and forced as new head on this PR's branch.

BastienLQ commented:
When a text file is sent, message is displayed as a file for the sender, but as a text MMS (with encoding issues) for the recipient.

I made some fixes in 633b910 / [`file-transfert`](https://github.com/SMSSecure/SMSSecure/commits/file-transfert) (indentation and assets). Can you base your modifications on this commit? I can also add you in the SMSSecure organization if you want.

vdveer commented:
That would safe me soms rebasing, so that would Be great. I will look into the text file issues tonight or the weekend. I had a merge conflict there with upstream changes, so I'll do some more testing.
